kasa/klapprotocol.py

```diff
@@ -0,0 +1,487 @@
+
+Encryption/Decryption methods based on the works of
+Simon Wilkinson and Chris Weeldon
+
+Klap devices that have never been connected to the kasa
+cloud should work with blank credentials.
+Devices that have been connected to the kasa cloud will
+switch intermittently between the users cloud credentials
+and default kasa credentials that are hardcoded.
+This appears to be an issue with the devices.
+
+The protocol works by doing a two stage handshake to obtain
+and encryption key and session id cookie.
+
+Authentication uses an auth_hash which is
+md5(md5(username),md5(password))
+
+handshake1: client sends a random 16 byte local_seed to the
+device and receives a random 16 bytes remote_seed, followed
+by sha256(local_seed + auth_hash).  It also returns a
+TP_SESSIONID in the cookie header.  This implementation
+then checks this value against the possible auth_hashes
+described above (user cloud, kasa hardcoded, blank).  If it
+finds a match it moves onto handshake2
+
+handshake2: client sends sha25(remote_seed + auth_hash) to
+the device along with the TP_SESSIONID.  Device responds with
+200 if succesful.  It generally will be because this
+implemenation checks the auth_hash it recevied during handshake1
+
+encryption: local_seed, remote_seed and auth_hash are now used
+for encryption.  The last 4 bytes of the initialisation vector
+are used as a sequence number that increments every time the
+client calls encrypt and this sequence number is sent as a
+url parameter to the device along with the encrypted payload
+
+https://gist.github.com/chriswheeldon/3b17d974db3817613c69191c0480fe55
+https://github.com/python-kasa/python-kasa/pull/117
+
+"""
+
+import asyncio
+import datetime
+import hashlib
+import logging
+import secrets
+import time
+from pprint import pformat as pf
+from typing import Any, Dict, Optional, Tuple, Union
+
+import httpx
+from cryptography.hazmat.primitives import hashes, padding
+from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
+
+from .credentials import Credentials
+from .exceptions import AuthenticationException, SmartDeviceException
+from .json import dumps as json_dumps
+from .json import loads as json_loads
+from .protocol import TPLinkProtocol
+
+_LOGGER = logging.getLogger(__name__)
+logging.getLogger("httpx").propagate = False
+
+
+class TPLinkKlap(TPLinkProtocol):
+    """Implementation of the KLAP encryption protocol.
+
+    KLAP is the name used in device discovery for TP-Link's new encryption
+    protocol, used by newer firmware versions.
+    """
+
+    DEFAULT_PORT = 80
+    DEFAULT_TIMEOUT = 5
+    DISCOVERY_QUERY = {"system": {"get_sysinfo": None}}
+    KASA_SETUP_EMAIL = "kasa@tp-link.net"
+    KASA_SETUP_PASSWORD = "kasaSetup"  # noqa: S105
+    SESSION_COOKIE_NAME = "TP_SESSIONID"
+
+    def __init__(
+        self,
+        host: str,
+        credentials: Optional[Credentials] = None,
+        *,
+        timeout: Optional[int] = None,
```

Comment:
```suggestion
        *,
        credentials: Optional[Credentials] = None,
        timeout: Optional[int] = None,
```
Let's enforce the design decision of having creds as a kwarg-only also for the protocol class.

Reply:
Done